File: python/algo/202606/algoiv04.09.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def BSTSequences(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        if root is None: return [[]]
        # 洗牌合并，保持左右子树的相对顺序不变
        def weave(a: List[int], b: List[int]) -> List[List[int]]:
            res = []
            path = []
            def backtrack(i, j):
                if i == len(a) and j == len(b):
                    res.append(path.copy())
                    return
                if i < len(a):
                    path.append(a[i])
                    backtrack(i+1, j)
                    path.pop()
                if j < len(b):
                    path.append(b[j])
                    backtrack(i, j+1)
                    path.pop()
            backtrack(0, 0)
            return res
        
        left = self.BSTSequences(root.left)
        right = self.BSTSequences(root.right)
        
        ans = []

        for l in left:
            for r in right:
                for mixed in weave(l, r):
                    ans.append([root.val] + mixed)

        return ans

File: cplus/leetcode/202606/algoiv04.09.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> BSTSequences(TreeNode* root) {
        if (root == nullptr) return {{}};

        auto weave = [](vector<int>& a, vector<int> b) -> vector<vector<int>> {
            vector<vector<int>> res;
            vector<int> path;
            auto backtrack = [&](this auto&& bk, int i, int j) -> void {
                if (i == a.size() && j == b.size()) res.emplace_back(path);

                if (i < a.size()) {
                    path.emplace_back(a[i]);
                    bk(i+1, j);
                    path.pop_back();
                }
                if (j < b.size()) {
                    path.emplace_back(b[j]);
                    bk(i, j+1);
                    path.pop_back();
                }
            };

            backtrack(0, 0);
            return res;
        };

        auto left = BSTSequences(root->left);
        auto right = BSTSequences(root->right);

        vector<vector<int>> ans;

        for (auto l : left) for (auto r: right) {
            auto ws = weave(l, r);
            for (auto & w: ws) {
                vector<int> res = {root->val};
                res.insert(res.end(), w.begin(), w.end());
                ans.emplace_back(res);
            }
        }

        return ans;
    }
};